sdk/search/azure-search/src/main/java/com/azure/search/SearchIndexAsyncClient.java

```diff
@@ -0,0 +1,732 @@
+     */
+    private final String endpoint;
+
+    /**
+     * The name of the Azure Cognitive Search index.
+     */
+    private final String indexName;
+
+    /**
+     * The logger to be used
+     */
+    private final ClientLogger logger = new ClientLogger(SearchIndexAsyncClient.class);
+
+    /**
+     * The underlying AutoRest client used to interact with the Azure Cognitive Search service
+     */
+    private final SearchIndexRestClientImpl restClient;
+
+    /**
+     * The pipeline that powers this client.
+     */
+    private final HttpPipeline httpPipeline;
+
+    /**
+     * Package private constructor to be used by {@link SearchIndexClientBuilder}
+     */
+    SearchIndexAsyncClient(String endpoint, String indexName, SearchServiceVersion apiVersion,
+        HttpPipeline httpPipeline) {
+
+        SearchServiceUrlParts parts = SearchServiceUrlParser.parseServiceUrlParts(endpoint);
+
+        if (StringUtils.isBlank(indexName)) {
+            throw logger.logExceptionAsError(new IllegalArgumentException("Invalid indexName"));
+        }
+        if (apiVersion == null) {
+            throw logger.logExceptionAsError(new IllegalArgumentException("Invalid apiVersion"));
+        }
+        if (httpPipeline == null) {
+            throw logger.logExceptionAsError(new IllegalArgumentException("Invalid httpPipeline"));
+        }
+
+        this.endpoint = endpoint;
+        this.indexName = indexName;
+        this.apiVersion = apiVersion;
+        this.httpPipeline = httpPipeline;
+
+        restClient = new SearchIndexRestClientBuilder()
+            .searchServiceName(parts.serviceName)
+            .indexName(indexName)
+            .searchDnsSuffix(parts.dnsSuffix)
+            .apiVersion(apiVersion.getVersion())
+            .pipeline(httpPipeline)
+            .serializer(SERIALIZER)
+            .build();
+    }
+
+    /**
+     * Gets the name of the Azure Cognitive Search index.
+     *
+     * @return the indexName value.
+     */
+    public String getIndexName() {
+        return this.indexName;
+    }
+
+    /**
+     * Gets the {@link HttpPipeline} powering this client.
+     *
+     * @return the pipeline.
+     */
+    HttpPipeline getHttpPipeline() {
+        return this.httpPipeline;
+    }
+
+    /**
+     * Uploads a collection of documents to the target index
+     *
+     * @param documents collection of documents to upload to the target Index.
+     * @return document index result.
+     */
+    public Mono<IndexDocumentsResult> uploadDocuments(Iterable<?> documents) {
+        return this.uploadDocumentsWithResponse(documents)
+            .map(Response::getValue);
+    }
+
+    /**
+     * Uploads a collection of documents to the target index
+     *
+     * @param documents collection of documents to upload to the target Index.
+     * @return response containing the document index result.
+     */
+    public Mono<Response<IndexDocumentsResult>> uploadDocumentsWithResponse(Iterable<?> documents) {
```

Comment:
@srnagar I can speculate on why this choice was made in this case based on my knowledge of the .NET SDK. The number of items being uploaded could potentially be large, and we wouldn't want to force users to reify all the items in memory at once (at least, more than is necessary to serialize the request body). If I understand correctly, using `Iterable` will allow customers to reify actions lazily. At some point in the future, we may change the underlying REST API to be a streaming API, and this choice would potentially allow us to take advantage of that without breaking changes.

I'm going to assume this isn't blocking the PR unless you tell me otherwise.